### PR TITLE
mgr/dashboard: introduce server side pagination for osds

### PR DESCRIPTION
**Pull request by nizamial09**
Fixes: https://tracker.ceph.com/issues/56511

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by nizamial09** [comment]
a draft, mainly to clean the code and fix any test failures.

File: src/pybind/mgr/dashboard/controllers/osd.py
Comment (epuertat) [inline src/pybind/mgr/dashboard/controllers/osd.py:128]: We should probably work with the Core team (CC @rzarzynski) to extend the underlying call, `mgr.get('osd_map')` (and others with potential scalability issues) with a common interface to allow random access to lists, and, perhaps, search/sorting capabilities:
```suggestion
        all_osds = self.get_osd_map(offset=int(offset), limit=int(limit), sort=sort, ... )
```
Long time ago, we identified that the process of converting internal Ceph arrays to Python objects was heavily impacting mgr scalability. However, Dashboard and other modules usually require a small subset of data than the one provided by the `get_osd_map` call, which ultimately means a waste of resources.

This should be made available to other ceph-mgr API methods like `mgr.get('osd_stats')`, `mgt.get('osd_map_tree')`, etc.

The approach could involve extending `OSDMap::dump(Formatter *f, CephContext *cct)` to support some iterator/slice semantics (start/offset, end/limit, sort, etc). Instead of altering the `dump()` itself, we might add the pagination to the `Formatter` (this would be consistent with the similar approach of C++ `ranges`/`views`, separating presentation from data holding).

---

**Alternatively, we could follow the opposite approach** (proposed by @tchaikov) of creating smart CPython classes, like the `OsdMap`:
https://github.com/ceph/ceph/blob/17dd1a270d3adeea850f0af4d6bc0ee207fab5dd/src/pybind/mgr/mgr_module.py#L201-L267

However, that's more challenging as it requires both expertise on CPython and C++ .

File: src/pybind/mgr/dashboard/openapi.yaml
Comment (nizamial09) [reply src/pybind/mgr/dashboard/openapi.yaml:8943]: i can open a tracker to track it differently. I suspect this will modify the entire openapi spec. Might be good to separate it.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts:435]: Can't do it. We get the table ccontext for this component within this file because that's where the table is rendered. And we need it increase/decrease the limit of items in a request.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts:441]: I'd be happy if this is dealt separately mainly to ease my backport process and avoid regressions (I am dealing with far too many this 2 weeks)

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/osd.service.ts
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/osd.service.ts]: done, thank you

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/osd.service.ts
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/osd.service.ts]: done as well. Thanks

File: src/pybind/mgr/dashboard/controllers/osd.py
Comment (nizamial09) [reply src/pybind/mgr/dashboard/controllers/osd.py:128]: yeah, this would actually improve dashboard performance by a lot..

**Comment by nizamial09** [comment]
jenkins retest this please

**Review by epuertat — APPROVED** [review]
LGTM! Thanks!

**Comment by nizamial09** [comment]
fixed make check failures and pushed again..

File: src/pybind/mgr/dashboard/frontend/src/app/shared/models/osd.model.ts
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/models/osd.model.ts:2]: why do we have all field optional ?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts:433]: why this "() => undefined", I am not following, sorry.

**Review by afreen23 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/controllers/osd.py
Comment (nizamial09) [reply src/pybind/mgr/dashboard/controllers/osd.py]: thanks @afreen23! applied your suggestion and also added a try-catch block for catching in type conversion failures. I was going to use `.isdigit()` but thought it might be better to explicitly catch the error and raise it.

Btw, the keys are always the same as the id since that is what being set in the [get_osd_map](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/controllers/osd.py#L206) method.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/models/osd.model.ts
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/models/osd.model.ts]: i made some of the optional and some of them required now.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts:433]: i just copied it from other parts but its basically passing an empty arrow function to the `CdTableFetchDataContext` because its constructor expects an error fn.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts:436]: can we add type for `resp`  here instead of `any` ?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts
Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts:433]: I see this is only for error and to be called from `fetchData()` and that we do call it at [some places.](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/services.component.ts#L247)

```
/**
   * The function that should be called from within the error handler
   * of the 'fetchData' function to display the error panel and to
   * reset the data table to the correct state.
   */
  error: Function;
  ```
Right now I dont see any instances of a custom error function in code.
As almost everywhere we require this to be unset `() => undefined`.

So we may want to revisit and think if its still required, only [Table Component ](https://github.com/ceph/ceph/blob/398563d959f96017fda47be5fac9d0a84c84ec1e/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts#L844)sets the context with error function, so perhaps we can bring that here in `CdTableFetchDataContext` and avoid sending it in params.

Not blocking this PR on this one, but since it came up here, hence commented.

But for this PR and also at other places may be we can just do if(context) and then proceed, why need a dummy context ?

**Review by afreen23 — APPROVED** [review]
Thanks looks good, just two minors, and  not holding this PR on that.
But expect some comments/talks :)

File: src/pybind/mgr/dashboard/frontend/src/app/shared/models/osd.model.ts
Comment (afreen23) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/models/osd.model.ts]: Okay, and what is the deciding criteria ? Or if we can have some comment here for future, will help, thanks!

**Comment by afreen23** [comment]
jenkins test api

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts:433]: if you check it in the osd-list.component, the `getOsdList()` is called in other places possibly to refresh the table after editing or updating some flags and during those calls, the context are not set. so if we just get the list only when we have `context` we'll be looking at an outdated data for these calls which is not what we need here (and certainly will be a regression).

maybe we look into making that error function an optional parameter in the constructor so we don't need to put undefined everytime, but shall we do it separately?

Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts]: for now I made the `error` optional in the `CdTableFetchDataContext` and removed the `undefined` function.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/models/osd.model.ts
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/models/osd.model.ts]: just added a comment there in the file. But there is not a definite schema in the back-end so its pretty difficult to understand what are all we need in the UI. To avoid any regressive behavior I just went with what I saw in some of the test cases as the deciding factor for making some items optional and non-optional.

Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/models/osd.model.ts]: We should def type things at the back-end and use some tooling to dynamically generate front-end models from that (Python typing -> JSONSchema -> Typescript model).

**Comment by afreen23** [comment]
jenkins retest this please

**Pull request merged**
